contracts/DecentHats.sol

```diff
@@ -0,0 +1,485 @@
+                toggle,
+                _hat.isMutable,
+                _hat.imageURI
+            );
+    }
+
+    function _createAccount(
+        IERC6551Registry _registry,
+        address _hatsAccountImplementation,
+        address protocolAddress,
+        uint256 hatId
+    ) internal returns (address) {
+        return
+            _registry.createAccount(
+                _hatsAccountImplementation,
+                SALT,
+                block.chainid,
+                protocolAddress,
+                hatId
+            );
+    }
+
+    function _createTopHatAndAccount(
+        IHats _hatsProtocol,
+        string memory _topHatDetails,
+        string memory _topHatImageURI,
+        IERC6551Registry _registry,
+        address _hatsAccountImplementation
+    ) internal returns (uint256 topHatId, address topHatAccount) {
+        topHatId = _hatsProtocol.mintTopHat(
+            address(this),
+            _topHatDetails,
+            _topHatImageURI
+        );
+
+        topHatAccount = _createAccount(
+            _registry,
+            _hatsAccountImplementation,
+            address(_hatsProtocol),
+            topHatId
+        );
+    }
+
+    function _createHatAndAccountAndMintAndStreams(
+        IHats hatsProtocol,
+        IERC6551Registry registry,
+        address topHatAccount,
+        address hatsAccountImplementation,
+        uint256 adminHatId,
+        Hat calldata hat
+    ) internal returns (uint256 hatId, address accountAddress) {
+        hatId = _createHat(
+            hatsProtocol,
+            adminHatId,
+            hat,
+            topHatAccount,
+            topHatAccount
+        );
+        accountAddress = _createAccount(
+            registry,
+            hatsAccountImplementation,
+            address(hatsProtocol),
+            hatId
+        );
+
+        hatsProtocol.mintHat(hatId, hat.wearer);
+
+        for (uint256 i = 0; i < hat.sablierParams.length; ) {
+            _createSablierStream(hat.sablierParams[i], accountAddress);
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    function _createTermedHatAndAccountAndMintAndStreams(
+        IHats hatsProtocol,
+        address topHatAccount,
+        address eligibilityAddress,
+        uint256 adminHatId,
+        Hat calldata hat
+    ) internal {
+        require(
+            hat.termedParams.length == 1,
+            "DecentHats: termedParams length must be 1"
+        );
+        require(
+            hat.termedParams[0].nominatedWearers.length == 1,
+            "DecentHats: nominatedWearers length must be 1"
+        );
+        uint256 hatId = _createHat(
+            hatsProtocol,
+            adminHatId,
+            hat,
+            topHatAccount,
+            eligibilityAddress
+        );
+
+        IHatsElectionEligibility(eligibilityAddress).elect(
+            hat.termedParams[0].termEndDateTs,
+            hat.termedParams[0].nominatedWearers
+        );
```

Comment:
Nitpick, flip the order of these two statements.

Reply:
?

Reply:
I guess it doesn't really matter, but it seem to make sense to create the hat then elect?

Reply:
lol yeah, it's just because the `hatId` variable isn't used in the `elect` call, but it's used in the line below that. My intention with the nitpick comment was just to keep relevant code close to each other.